presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java

```diff
@@ -0,0 +1,189 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.Page;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import javax.annotation.Nullable;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+import static java.util.Objects.requireNonNull;
+
+public class MergeSortProcessor
+{
+    private static final ListenableFuture<?> NOT_BLOCKED = immediateFuture(null);
+
+    private final MergeSortComparator comparator;
+    private final List<MergeSource> mergeSources;
+
+    private ListenableFuture<?> blocked;
+
+    public MergeSortProcessor(MergeSortComparator comparator, List<? extends PageSupplier> pageSuppliers)
+    {
+        this.comparator = requireNonNull(comparator, "comparator is null");
+        requireNonNull(pageSuppliers, "pageSuppliers is null");
+        ImmutableList.Builder<MergeSource> mergeSources = ImmutableList.builder();
+        for (PageSupplier pageSupplier : pageSuppliers) {
+            mergeSources.add(new MergeSource(pageSupplier));
+        }
+        this.mergeSources = mergeSources.build();
+    }
+
+    @Nullable
+    public PageWithPosition poll()
```

Comment:
> I agree that it's weird, but is it worth a data copy to make the code cleaner?

You don't need to copy the data, just the create new `PageWithPosition` with incremented position variable. Sure, you would need to create new `PageWithPosition` for each row, but I think it's worth it for the sake of simplicity and cleanness. 

See how `MergeHashSort` is doing that with iterator approach: `MergeHashSort.SingleChannelPagePositions` and `MergeHashSort.PageRewriteIterator`. It uses Guavas `Iterators.mergeSorted` to have log number of comparisons instead of linear